Use Travis CI to build MinGW packages

liushuyu opened the pull request:
As it was said before, this is precedent to the PR #3649 

This one maybe controversial, so I expect it will take time to be reviewed and merged.

There's a workaround in the build script, the reason it was here was that `libressl` module wasn't up-to-date in `citra` main repository.

Libraries used in this build:

Lib|Version
-----|-----
Qt 5 |5.9.6
SDL2 | 2.0.7
libsamplerate | 0.1.9

Test build download link is shown here: https://travis-ci.org/liushuyu/citra/jobs/373341674#L4134

The download speed maybe very slow, and it will be expired after 14 days (2018-05-13)

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/citra-emu/citra/3706)
<!-- Reviewable:end -->

liushuyu commented:
And maybe after this PR is merged, we can remove the MinGW build from Appveyor to cut overall CI time down to around 20 minutes.

Another side effect observed was that UI components now follows native components theme, the binary size shrieked to 15 MB compressed, 65.8 MB uncompressed

jroweboy commented on .travis/linux-mingw/docker.sh:
```diff
@@ -0,0 +1,52 @@
+apt-get update
+apt-get install -y software-properties-common wget git python3-pip ccache g++-mingw-w64-x86-64 gcc-mingw-w64-x86-64 mingw-w64-tools
+
+# temporary workaround
```
we did merge the update pr to the externals repo. If we need to update it again, feel free to send a pr and i'll make sure it gets merged soon :)

liushuyu replied:
The problem is the main repository is unaware of the changes of the submodule.
You need to update the submodule in the main repository and commit the change. (b/c the submodule works like a pointer, you need to update the address of that pointer)

jroweboy replied:
yes, and thats something I was expecting you would do as part of this pull request

jroweboy commented on .travis/linux-mingw/docker.sh:
```diff
@@ -0,0 +1,52 @@
+# !temporary workaround
+
+add-apt-repository ppa:tobydox/mingw-w64 -y
+# HACK: the repository does not contain necessary packages for 18.04, we'll use packages for 17.10 for now
```
i take it that this works, but its kinda surprising to me that it does :)

liushuyu replied:
The directory structures for MinGW toolchain stayed the same and they are not binaries for the host system, so that's kinda easy to explain why this worked

jroweboy commented on .travis/linux-mingw/docker.sh:
```diff
@@ -0,0 +1,52 @@
+export PATH=/citra/cmake/cmake-3.10.1-Linux-x86_64/bin:$PATH
+
+# fix a problem in current MinGW headers
+wget -q https://github.com/Alexpux/mingw-w64/raw/master/mingw-w64-headers/crt/errno.h -O /usr/x86_64-w64-mingw32/include/errno.h
```
i'd rather freeze this to a specific commit instead of master just in case it changes before appveyor decides to update msys2 again.

liushuyu replied:
I will pinpoint a specified commit

jroweboy commented on .travis/linux-mingw/docker.sh:
```diff
@@ -0,0 +1,52 @@
+
+for i in package/*.exe; do
+  # we need to process pdb here, however, cv2pdb
+  # does not work here, so we just simply strip all the debug symbols
```
this seems a little regrettable, as we were planning on using the debug symbols with minidump to get stacktraces. i don't think we should require symbols, but if someone has any good ideas on how to address this, i'd like to hear them :)

liushuyu replied:
I will try to see if `cv2pdb` can be ported to the build environment.

liushuyu replied:
Okay, `cv2pdb` contains enormous Windows specific code and even MS Assembly.

>but if someone has any good ideas on how to address this, i'd like to hear them :)

Well, I don't know if you are interested in Google Breakpad the crash reporting system used in Chromium/Chrome

jroweboy commented on .travis/linux-mingw/docker.sh:
```diff
@@ -0,0 +1,52 @@
+cd ..
+mkdir package
+
+QT_PLATFORM_DLL_PATH='/usr/x86_64-w64-mingw32/lib/qt5/plugins/platforms/'
```
not sure right now, but just double checking that the platforms directory for this qt only has qwindows.dll and none of the other platform plugins, correct?

liushuyu replied:
There are several files under `platform` directory. `qwindows.dll` and `qminimal.dll` to name a few.

jroweboy replied:
as i recall, we only need qwindows.dll at the moment, so maybe just hard code that.

wwylele replied:
Though #3566 is probably introducing some new platform plugins.

zhaowenlan1779 replied:
Yes, will also need mediaservice

liushuyu replied:
>>Though #3566 is probably introducing some new platform plugins.

>No it only adds mediaservice plugins

Fortunately, the PPA I used contains QtMultiMedia libraries for MinGW, it should not be a problem.

I wonder what else need to be done to push forward this PR?

jroweboy commented on .travis/linux-mingw/scan_dll.py:
```diff
@@ -0,0 +1,100 @@
+    results = []
+    pe = pefile.PE(file_name, fast_load=True)
+    pe.parse_data_directories()
+
```
i'm not too familiar with the PE library used in python here, but is there a way we can replicate how we are detecting system dlls like we currently do here: https://github.com/citra-emu/citra/blob/9c65a45358f0d088fcb78ca7f7ade3772099d7f8/.appveyor/FindDependencies.ps1#L27 this method of hardcoded system dlls and checking for 32.dll seems just a little fragile.

liushuyu replied:
This PE library is only a static analysis library. Remember that Appveyor is using a Windows (R) Operating system and `dumpbin` uses dynamic analysis. So the method described in `citra/.appveyor/FindDependencies.ps1`does not work here since it does not have the concept of `WIN32 SystemRoot` or something like that.

liushuyu replied:
>this method of hardcoded system dlls and checking for 32.dll seems just a little fragile.

As mentioned before, but just to clarify a bit more, since the host is not Windows, obviously these dlls are not coming from `C:\` of course.

jroweboy commented:
https://github.com/citra-emu/citra/pull/3566/files#diff-180360612c6b8c4ed830919bbb4dd459R132
No it only adds mediaservice plugins

On Tue, May 1, 2018 at 3:37 PM, Weiyi Wang <notifications@github.com> wrote:

> *@wwylele* commented on this pull request.
> ------------------------------
>
> In .travis/linux-mingw/docker.sh
> <https://github.com/citra-emu/citra/pull/3706#discussion_r185341345>:
>
> > +
> +# fix a problem in current MinGW headers
> +wget -q https://github.com/Alexpux/mingw-w64/raw/master/mingw-w64-headers/crt/errno.h -O /usr/x86_64-w64-mingw32/include/errno.h
> +
> +mkdir build && cd build
> +cmake .. -DCMAKE_TOOLCHAIN_FILE="$(pwd)/../CMakeModules/MinGWCross.cmake" -DUSE_CCACHE=ON -DCMAKE_BUILD_TYPE=Release -DENABLE_QT_TRANSLATION=ON -DCITRA_ENABLE_COMPATIBILITY_REPORTING=${ENABLE_COMPATIBILITY_REPORTING:-"OFF"}
> +make -j4
> +
> +echo "Tests skipped"
> +#ctest -VV -C Release
> +
> +echo 'Prepare binaries...'
> +cd ..
> +mkdir package
> +
> +QT_PLATFORM_DLL_PATH='/usr/x86_64-w64-mingw32/lib/qt5/plugins/platforms/'
>
> Though #3566 <https://github.com/citra-emu/citra/pull/3566> is probably
> introducing some new platform plugins.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/citra-emu/citra/pull/3706#discussion_r185341345>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AA6Iw8mwq1yGO4RFUuhucxm2-7MiwbZQks5tuNWAgaJpZM4Ttl17>
> .
>

liushuyu commented:
To whom may concerned, this PR also had addressed the problem in #3716.
The Qt in PPA has statically linked the `dll` symbols inside the modules.

jroweboy commented on .travis/linux-mingw/upload.sh:
```diff
@@ -0,0 +1,13 @@
+
+mkdir "$REV_NAME"
+# get around the permission issues
+cp -r package/* "$REV_NAME"
```
this root package should be named `head-mingw/` instead of `package/` in order to replace the existing mingw uploads.

liushuyu replied:
The packaged directory is actually `$REV_NAME`. The `package` directory is a temporary storage, since the files and directory itself is under `root` user, we can't `mv` it or change files in it.

jroweboy replied:
Ah my mistake, I thought this was getting the repo tag name "head" "canary" or "nightly" but it looks like thisi s just for the name of the zip.

liushuyu commented:
Very well, I'll address your comments later today.

jroweboy commented:
Since this conflicts with the frontend camera support, I'm planning on merging that one first and rebasing this one off that. My idea is that after merging camera and rebasing this, we can tag this for canary to make sure that everything will still work with the installer and that the transition will be smooth. Are you fine with this?

jroweboy commented:
@liushuyu qt camera has been merged. When you rebase and get it working in travis mingw, lemme know and i'll tag it for canary

liushuyu commented:
>@liushuyu qt camera has been merged. 

Wow, I didn't expect it to be merged this fast.

>When you rebase and get it working in travis mingw, lemme know and i'll tag it for canary

I will finish it probably today if nothing goes south. :smile:

liushuyu commented:
>Was this intending on replacing the mingw build entirely? If so we should also do something to disable the appveyor build https://github.com/citra-emu/citra/blob/master/appveyor.yml#L16 maybe comment out mingw (or if you are feeling adventurous, rip it out of the file entirely)

Removed in commit b0b9ba8 . We are using a VCS, so I don't comment out things. 

Just remember not to do a squash merge and everyone will be happy :smile_cat:

j-selby commented on .travis.yml:
```diff
@@ -49,6 +49,19 @@ matrix:
       cache: ccache
       script: "./.travis/linux-frozen/build.sh"
+    - os: linux
+      env: NAME="linux build (MinGW)"
```
Feel like the names of the other builds ("linux build") should also be changed to be differentiated, if you are going to be naming them all linux build. Can I suggest that you drop the Linux build stuff (Linux can already be implied) and instead name this something like "cross-compile build (MinGW)"?

liushuyu replied:
How about `MinGW build` straight?

j-selby replied:
sure, sounds good!

wwylele commented on .travis/linux-mingw/docker.sh:
```diff
@@ -0,0 +1,43 @@
+wget -q https://github.com/Alexpux/mingw-w64/raw/d0d7f784833bbb0b2d279310ddc6afb52fe47a46/mingw-w64-headers/crt/errno.h -O /usr/x86_64-w64-mingw32/include/errno.h
+
+mkdir build && cd build
+cmake .. -DCMAKE_TOOLCHAIN_FILE="$(pwd)/../CMakeModules/MinGWCross.cmake" -DUSE_CCACHE=ON -DCMAKE_BUILD_TYPE=Release -DENABLE_QT_TRANSLATION=ON -DCITRA_ENABLE_COMPATIBILITY_REPORTING=${ENABLE_COMPATIBILITY_REPORTING:-"OFF"}
```
Need to add flag `-DENABLE_COMPATIBILITY_LIST_DOWNLOAD=ON` do include the compatibility list

liushuyu replied:
>Need to add flag -DENABLE_COMPATIBILITY_LIST_DOWNLOAD=ON do include the compatibility list

I don't know this option existed... I'll add that now, the options are copypasta from other build jobs.

wwylele commented:
Just put this here so that we don't lose record:

We have got some report saying that the travis build had worse performance than the appveyor build (of the same canary version, of course). We still don't have idea why it is like this.

Leo626 commented:
I noticed now that the system camera shows a green screen with this. A lot of dll files are missing including the mediaservice and styles folder. Not sure if that was the cause of this.

Canary-823fe4c (496)

![explorer_2018-05-16_22-48-00](https://user-images.githubusercontent.com/17358613/40154353-5264586c-5964-11e8-84cc-2477957400c0.png)

Canary-e9bd59d (490)

![explorer_2018-05-16_22-50-01](https://user-images.githubusercontent.com/17358613/40154361-590559aa-5964-11e8-9f57-33714a811451.png)

liushuyu commented:
>I noticed now that the system camera shows a green screen with this. A lot of dll files are missing including the mediaservice and styles folder. Not sure if that was the cause of this.

@Leo626 Thanks for figuring out the problem, the mediaservice plugins are now copied. The other DLLs you spot missing were intentional move. These plugins functionalities are now included in Qt5* DLLs.

zhaowenlan1779 commented:
latest canary
![image](https://user-images.githubusercontent.com/21307832/40268249-a8d47c4c-5b9c-11e8-8f67-ebeb61fcb4dd.png)

I guess this was "Citra Canary" before?

And this one?
![image](https://user-images.githubusercontent.com/21307832/40268257-be3bb46a-5b9c-11e8-8db8-eb8c80e52e76.png)

And the citra log.

zhaowenlan1779 commented:
@liushuyu 
> Another side effect observed was that UI components now follows native components theme, the binary size shrieked to 15 MB compressed, 65.8 MB uncompressed

Note that's because you are using Qt5.9.6 (<Qt5.10).
It's after Qt5.10 that they got a styles folder.

jroweboy commented:
scan_dll just had an error where it couldn't find avrt.dll? https://travis-ci.org/citra-emu/citra-canary/jobs/383713359#L3431

liushuyu commented:
>scan_dll just had an error where it couldn't find avrt.dll? https://travis-ci.org/citra-emu/citra-canary/jobs/383713359#L3431
>
>Note: this started happening when we added the cubeb PR to canary

Fixed. Now the script will no longer raise error when DLL is missing, it will only warn you.

>I guess this was "Citra Canary" before?
>
>And this one?

@zhaowenlan1779  I have no idea what variable or option can control the name? Care to elaborate? Thanks.

jroweboy commented:
In common/cmakelists is where it pulls the repo name from appveyor and
Travis to set up the scm_rev.h file with the name of the build

On Fri, May 25, 2018, 8:32 PM liushuyu <notifications@github.com> wrote:

> scan_dll just had an error where it couldn't find avrt.dll?
> https://travis-ci.org/citra-emu/citra-canary/jobs/383713359#L3431
>
> Note: this started happening when we added the cubeb PR to canary
>
> Fixed. Now the script will no longer raise error when DLL is missing, it
> will only warn you.
>
> I guess this was "Citra Canary" before?
>
> And this one?
>
> @zhaowenlan1779 <https://github.com/zhaowenlan1779> I have no idea what
> variable or option can control the name? Care to elaborate? Thanks.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/citra-emu/citra/pull/3706#issuecomment-392229524>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AA6Iw98Y4yNFz4G2cZOTA4aGuJNUeiwCks5t2L7YgaJpZM4Ttl17>
> .
>

liushuyu commented:
>In common/cmakelists is where it pulls the repo name from appveyor and
Travis to set up the scm_rev.h file with the name of the build

I have checked the CMake file and I can't see there's anything abnormal. Since MinGW also uses Travis CI and build from the canary repository, then why it didn't get the `canary` name?

wwylele commented:
Temporarily untagging this because I want to get a working canary build.

liushuyu commented:
@zhaowenlan1779 Your concern is being addressed in #3825

jroweboy commented:
@liushuyu please update this with the fix for the env vars and now that we have cubeb audio (and don't have to rely on sdl2 audio) i think we are fine to merge this.

liushuyu commented:
>@liushuyu please update this with the fix for the env vars and now that we have cubeb audio (and don't have to rely on sdl2 audio)

I'll sure rebase the PR with new commits. 

> i think we are fine to merge this.

Please don't merge yet, I would liek to wait for PPA owner to update the packages.

liushuyu commented:
>@liushuyu please update this with the fix for the env vars and now that we have cubeb audio (and don't have to rely on sdl2 audio)

Successfully rebased. All commits have been added to the base.

I can see @zhaowenlan1779 and @jroweboy still have the unresolved review questions.
I think I have addressed the problems you raised before. Feel free to check if there are any other problems since this PR also affects self-hosted CI (if applicable) in the future.

liushuyu commented:
@zhaowenlan1779 Environment variables fix is here and you have reviewed this PR: 
#3825

wwylele commented on .travis/linux-mingw/build.sh:
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash -ex
+mkdir "$HOME/.ccache" || true
+docker run -e ENABLE_COMPATIBILITY_REPORTING -v $(pwd):/citra -v "$HOME/.ccache":/root/.ccache ubuntu:18.04 /bin/bash -ex /citra/.travis/linux-mingw/docker.sh
```
Shouldn't also add `--env-file .travis/common/travis-ci.env` here?

liushuyu replied:
Correct. I'll address this later.

liushuyu replied:
Fixed

zhaowenlan1779 commented on .travis/linux-mingw/docker.sh:
```diff
@@ -0,0 +1,46 @@
+
+mkdir package/platforms
+cp "${QT_PLATFORM_DLL_PATH}/qwindows.dll" package/platforms/
+cp -r "${QT_PLATFORM_DLL_PATH}/../mediaservice/" package/
```
Em. This copies all the dlls to the destination. But in fact that's not necessary. For example this includes a few `xxxxxxd.dll`, they are only used for Debug and is completely useless for Release builds.

zhaowenlan1779 commented on .travis/linux-mingw/docker.sh:
```diff
@@ -0,0 +1,46 @@
+
+mkdir package/platforms
+cp "${QT_PLATFORM_DLL_PATH}/qwindows.dll" package/platforms/
+cp -r "${QT_PLATFORM_DLL_PATH}/../mediaservice/" package/
```
Em. This copies all the dlls to the destination. But in fact that's not necessary. For example this includes a few xxxxxxd.dll, they are only used for Debug and is completely useless for Release builds.

liushuyu replied:
I'll take a look at this.

liushuyu replied:
@zhaowenlan1779 Corrected in 3a41f3a

liushuyu commented:
Okay... I decided not to continue waiting for upstream provider to update `SDL 2` since we have `cubeb` now. I'll still keep track of the update, it may take a while.

I'll address the last few issues here. 
And if everything looks good to you, then just merge.

jroweboy commented:
Lemme know when you get a chance to fix the last few outstanding issues, so we can get this merged

liushuyu commented:
And for #3891 I want to wait for THAT PR to be merged first and I'll match the changes in this PR.

jroweboy commented:
@liushuyu #3891 is now merged :)

liushuyu commented:
After this got merged, #3839 will be updated with changes in this PR

CeruleanSky commented:
The appveyor creates an archive that just has the pdb debug symbols for the canary builds, For example for build 632  [citra-windows-mingw-20180712-76a3229-debugsymbols.zip](https://ci.appveyor.com/project/citrabot/citra-canary/build/job/02i0x1v2lif6o4u4/artifacts).  Travis doesn't seem to have a [link for those](https://travis-ci.org/citra-emu/citra-canary/jobs/403156782).   I was wondering if this could be easily modified to either create the debug symbols artifact and make it downloadable from travis or the github page, duplicating the appveryor behavior.  That way it is easier to get better stack back traces, especially these versions that tend to be the most unstable and in need of the best diagnostics and testing.

jroweboy commented:
@CeruleanSky Appveyor will continue to be used for validating that builds are working properly from our two windows environments MSVC and MINGW64, and artifacts for debug symbols can still be found there and will still be there for the forseeable future. All this PR does is change which service uploads the build to github
